### PR TITLE
[ASA 9675] DAST rescan

### PR DESCRIPTION
**Pull request by vishalhcl-5960**
Implementation of rescan functionality for the dynamic scans.

### CODE REVIEW

File: src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java [hunk 1]
```diff
@@ -223,77 +223,151 @@ private static boolean hasEntitlement(String scanType, IAuthenticationProvider p
...
+        return null;
+    }
+
+    /**
+     * Fetch the details of all the executions of a scan.
+     *
+     * @param provider The IAuthenticationProvider for authentication.
+     * @return String.
+     */
+    public static String getA360Version(IAuthenticationProvider provider) {
```
Comment (mattmurp) [inline src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java:349]: There are a number of problems with this method:

1. The javadoc comment is not accurate.
2. The method name should just be getServiceVersion, not getA360Version.
3. We shouldn't need a valid token to check the service version.
4. We shouldn't need the authentication header.
5. You should use "body.getString()" as opposed to casting.
6. The comment in the catch block doesn't make sense.

File: src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java [hunk 2]
```diff
@@ -223,77 +223,151 @@ private static boolean hasEntitlement(String scanType, IAuthenticationProvider p
...
+    }
+
+    /**
+     * Fetch the details of all the executions of a scan.
+     *
+     * @param scanId The scanId to test
+     * @param provider The IAuthenticationProvider for authentication.
+     * @return JSONArray.
+     */
+    public static JSONArray getExecutionDetails(String scanId, IAuthenticationProvider provider) {
```
Comment (mattmurp) [inline src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java:318]: Why are there DAST specific checks here? The "IsValidForIncremental" property would only be true for DAST scans, but what if someone called this method for a SAST or SCA scan? This seems like a very specific method that is being added as a general utility method.

File: src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java [hunk 3]
```diff
@@ -223,77 +223,151 @@ private static boolean hasEntitlement(String scanType, IAuthenticationProvider p
...
-    public static void updateScanData(Map<String, String> params, String scanId, IAuthenticationProvider provider, IProgress progress) {
+    /**
+     * Fetch the detailed description of a scan.
+     *
+     * @param type The selected scan type
+     * @param scanId The scanId to test
+     * @param provider The IAuthenticationProvider for authentication.
+     * @return JSONObject.
+     */
+    public static JSONObject scanSpecificDetails(String type, String scanId, IAuthenticationProvider provider) {
```
Comment (mattmurp) [inline src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java:287]: The name of this method should be an action, like "getScanDetails()".

**Review by mattmurp — CHANGES_REQUESTED** [review]

File: src/main/java/com/hcl/appscan/sdk/utils/ArchiveUtil.java [hunk 4]
```diff
@@ -123,9 +123,6 @@ public static void zipFileOrFolder(File fileToZip, File zipFile) throws IOExcept
     }
 
     private static void zipFile(File fileToZip, String fileName, ZipOutputStream zipOut) throws IOException {
-        if (fileToZip.getName().startsWith(".")) {
```
[the remark this answers is not in the document]
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/sdk/utils/ArchiveUtil.java:126]: This change is being made to include the ".git" folder when creating a zip archive of the parent folder from a Git repository. 
Task details: https://jira02.hclpnp.com/browse/ASA-9830

File: src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java (hunk 3 above)
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java:287]: Sure, renaming the method to "getScanDetails()".

File: src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java (hunk 2 above)
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java:318]: This method is used to fetch the "executionIds" which helps to fill the base scan dropdown field of incremental scan. I am renaming this method to "getBaseScanDetails" for better understanding.

File: src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java (hunk 1 above)
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java]: Sure, Made the suggested changes.

File: src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java [hunk 5]
```diff
@@ -223,77 +223,141 @@ private static boolean hasEntitlement(String scanType, IAuthenticationProvider p
...
-    public static void updateScanData(Map<String, String> params, String scanId, IAuthenticationProvider provider, IProgress progress) {
+    /**
+     * Fetch the detailed description of a scan.
+     *
+     * @param type The selected scan type
+     * @param scanId The scanId to test
+     * @param provider The IAuthenticationProvider for authentication.
+     * @return JSONObject.
+     */
+    public static JSONObject getScanDetails(String type, String scanId, IAuthenticationProvider provider) {
```
Comment (mattmurp) [inline src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java:287]: What's the need for the addition of this method here when we already have the CloudScanServiceProvider that does the same thing?

File: src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java [hunk 6]
```diff
@@ -223,77 +223,141 @@ private static boolean hasEntitlement(String scanType, IAuthenticationProvider p
...
+    }
+
+    /**
+     * Fetch the details of all the executions of a scan.
+     *
+     * @param scanId The scanId to test
+     * @param provider The IAuthenticationProvider for authentication.
+     * @return JSONArray.
+     */
+    public static JSONArray getBaseScanDetails(String scanId, IAuthenticationProvider provider) {
```
Comment (mattmurp) [inline src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java:318]: This method would be better as a part of the IScanServiceProvider (i.e. CloudScanServiceProvider) as opposed to being added as a utility method here.

File: src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java [hunk 7]
```diff
@@ -223,77 +223,141 @@ private static boolean hasEntitlement(String scanType, IAuthenticationProvider p
...
-     * @param applicationId The applicationId to verify.
-     * @param type The scanType to verify.
+     * @param scanId The scanId of the scan whose configuration has to update.
+     * @param params The map of properties which has to update .
      * @param provider The IAuthenticationProvider for authentication.
-     * @return True if the scanId is valid. False is returned if the scanId is not valid, the request fails, or an exception occurs.
+     * @param progress The IProgress for setting the status messages.
      */
-    public static boolean isScanId(String scanId, String applicationId, String type, IAuthenticationProvider provider) {
+    public static void updateScanData(Map<String, String> params, String scanId, IAuthenticationProvider provider, IProgress progress) {
```
Comment (mattmurp) [inline src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java:246]: Similar to the other comments, it's not clear why this is being added as a utility method, as opposed to the IScanServiceProvider that performs related functions.

**Review by mattmurp — CHANGES_REQUESTED** [review]

File: src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java (hunk 5 above)
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java:287]: There are difference in the APIs. The method which is there in the CloudScanServiceProvider call the general API "GET /api/v4/Scans $filter=Id eq <scanId>" to fetch the scan details but new method of ServiceUtil calls the scan specific API "GET /api/v4/<scanType>/<scanId>".
For scanId validation, We want to get the value of the "gitRepoPlatform" parameter which is not present in the respeonse of general API.

File: src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java (hunk 7 above)
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java:246]: Okay, I will move it to the cloudScanServiceProvider class.

File: src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java (hunk 5 above)
Comment (vishalhcl-5960) [reply src/main/java/com/hcl/appscan/sdk/utils/ServiceUtil.java:258]: Matt, Should we move this new method into CloudScanServiceProvider class as it is perform related function?

File: src/main/java/com/hcl/appscan/sdk/scan/CloudScanServiceProvider.java [hunk 8]
```diff
@@ -269,4 +298,51 @@ private boolean verifyApplication(String appId) {
...
+				return (JSONArray) response.getResponseBodyAsJSON();
+			}
+		} catch (IOException | JSONException e) {
+			// Ignore and move on.
+		}
+
+		return null;
+	}
+
+	public void updateScanData(Map<String, String> params, String scanId, IAuthenticationProvider provider, IProgress progress) {
```
Comment (mattmurp) [inline src/main/java/com/hcl/appscan/sdk/scan/CloudScanServiceProvider.java:328]: Why does this method take an IAuthenticationProvider and an IProgress as parameters when there are already member variables of those types?

**Review by mattmurp — APPROVED** [review]

**Pull request merged**